Low-impact cross fixes

dtzWill opened the pull request:
###### Motivation for this change

Not exhaustive, but manageable chunk of fixes from #30882 (and a few local ones) that don't cause mass-rebuild.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bgamari committed:
xz: Don't run testsuite when cross-compiling

bgamari committed:
binutils: Add support for overriding default ld

bgamari committed:
krb5: Allow cross-compilation

bgamari committed:
libtasn1: Enable cross-compilation

bgamari committed:
autogen: Enable cross-compilation

bgamari committed:
pkgconfig: Enable cross-compilation

(cherry picked from commit 91fdb2f9caf5ce14acb87f3b877216aeb2590e7e)

bgamari committed:
xorg: Use build dependencies where appropriate

bgamari committed:
libuv: Disable tests when cross-compiling

dtzWill committed:
lua-5: isMingw -> isMinGW

dtzWill committed:
bsdiff: fix 'cc' usage

dtzWill committed:
at: cross-compilation fixes

* move bison/flex to nativeBuildInputs
* add patch to avoid use of undocumented glibc macro
* fix use of 'install -s' to avoid invoking 'strip' directly

Ericson2314 commented on pkgs/development/compilers/gcc/6/default.nix:
```diff
@@ -314,7 +314,44 @@ stdenv.mkDerivation ({
     export lib=$out;
   ''
-  ;
+  + stdenv.lib.optionalString (buildPlatform != hostPlatform)
```
Let's skip GCC for the moment

Ericson2314 commented on pkgs/development/compilers/llvm/4/llvm.nix:
```diff
@@ -27,6 +28,17 @@ let
   # Used when creating a version-suffixed symlink of libLLVM.dylib
   shortVersion = with stdenv.lib;
     concatStringsSep "." (take 2 (splitString "." release_version));
+
```
And also skip llvm

Ericson2314 commented on pkgs/development/libraries/libgpg-error/default.nix:
```diff
@@ -16,6 +16,7 @@ stdenv.mkDerivation rec {
 
   # If architecture-dependent MO files aren't available, they're generated
   # during build, so we need gettext for cross-builds.
+  crossAttrs.nativeBuildInputs = [ gettext buildPackages.stdenv.cc ];
```
Need to get rid of cross attrs and some other things. I can force push to fix or just skip for now.

Ericson2314 commented on pkgs/development/libraries/libssh2/default.nix:
```diff
@@ -29,6 +29,14 @@ stdenv.mkDerivation rec {
     '';
   };
 
+  # Hack: when cross-compiling we need to manually add rpaths to ensure that
```
No longer need this hack.

Ericson2314 reviewed (approved):
Good after those few things. Thanks!

dezgeg commented on pkgs/development/libraries/libuv/default.nix:
```diff
@@ -38,7 +38,7 @@ stdenv.mkDerivation rec {
   # For now the tests are just breaking large swaths of the nixpkgs binary cache for Darwin,
   # and I'd rather have everything else work at all than have stronger assurance here.
-  doCheck = !stdenv.isDarwin;
+  doCheck = !stdenv.isDarwin && stdenv.buildPlatform == stdenv.hostPlatform;
```
We shouldn't propagate this everywhere. Stdenv should just disable doCheck when cross compiling.

dezgeg commented on pkgs/tools/compression/xz/default.nix:
```diff
@@ -10,7 +10,7 @@ stdenv.mkDerivation rec {
   outputs = [ "bin" "dev" "out" "man" "doc" ];
 
-  doCheck = true;
+  doCheck = stdenv.hostPlatform == stdenv.buildPlatform;
```
Ditto.

dtzWill commented:
* Dropped gcc, llvm commits, per request.  Should they go into a separate PR?
* Dropping libgpg-error commit, I'm not sure what you'd like done.  Can you make a PR for this?
* Dropping libssh2 hack, glad it's not needed :).
* Rebased onto master since I'm force-pushing anyway

dtzWill commented:
@dezgeg would you mind if the stdenv doCheck fixes, this was discussed and the consensus seems to be that this is the way to go for now: https://github.com/NixOS/nixpkgs/pull/30883

As mentioned there, cleanup would be straightforward  if we find and agree on how to better handle this.

dezgeg commented:
No, I really think that pattern shouldn't spread to everywhere in nixpkgs.

dtzWill commented:
Dropping perl-cross, needs some work (not being critical, but it's just not incomplete and doesn't work yet).

Thought it did work, perhaps something it needs got lost in some shuffling, will pull it out.

Ericson2314 commented:
Ok the permenant solution I want for `doCheck` is a mass rebuild, but I can dig up a non-mass-rebuild rebuild for now.

dtzWill commented:
@Ericson2314 if you could revive that it'd be great!
